### Analysis the popularity of the free apps in the markets: Google Play and App Store

#### The rationale for this analysis

1. The customer of this analysis is a company developing free apps for both Android (Google Play market) and iOS (App Store).
2. Applications developed by the client are free - with all income comming from running the in-app advertisements.
3. As a result of the previous point the profit on application is directly dependent only on the number of users engaging inside of it with the served advertisements.
4. We are also interested only in applications which have support for English language.

#### The defined-upfront company's strategy for developping a profitable app is the following:

1. Build a minimal Android version of the app, and add it to Google Play.
2. If the app has a good response from users, we develop it further.
3. If the app is profitable after six months, we build an iOS version of the app and add it to the App Store.

#### Used datasets
1. [Google Play dataset](https://www.kaggle.com/datasets/lava18/google-play-store-apps)
2. [App Store dataset](https://www.kaggle.com/datasets/ramamet4/app-store-apple-data-set-10k-apps)


#### The research problem 

What are the characteristics of the applications attractive many users?

### Reading the data


In [1]:
from csv import reader

appstore_file_name = "AppleStore.csv"
playstore_file_name = "googleplaystore.csv"

def read_dataset(file_name):
    return list(reader(open(file_name, encoding='utf-8')))

def explore_data(dataset, start, end, rows_and_columns=False):
    dataset_slice = dataset[start:end]    
    for row in dataset_slice:
        print(row)
        print('\n') # adds a new (empty) line after each row

    if rows_and_columns:
        print('Number of rows:', len(dataset))
        print('Number of columns:', len(dataset[0]))

appstore_data = read_dataset(appstore_file_name)
playstore_data = read_dataset(playstore_file_name)

### App store dataset:

In [2]:
explore_data(appstore_data, 0, 1, True)

['id', 'track_name', 'size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic']


Number of rows: 7198
Number of columns: 16


In [3]:
explore_data(appstore_data, 1, 5, False)

['284882215', 'Facebook', '389879808', 'USD', '0.0', '2974676', '212', '3.5', '3.5', '95.0', '4+', 'Social Networking', '37', '1', '29', '1']


['389801252', 'Instagram', '113954816', 'USD', '0.0', '2161558', '1289', '4.5', '4.0', '10.23', '12+', 'Photo & Video', '37', '0', '29', '1']


['529479190', 'Clash of Clans', '116476928', 'USD', '0.0', '2130805', '579', '4.5', '4.5', '9.24.12', '9+', 'Games', '38', '5', '18', '1']


['420009108', 'Temple Run', '65921024', 'USD', '0.0', '1724546', '3842', '4.5', '4.0', '1.6.2', '9+', 'Games', '40', '5', '1', '1']




**Columns which might be useful:**
- **id** - App Id
- **track_name** - App Name
- **price** - Price amount (should be equal to zero)
- **rating_count_tot** - User Rating counts (for all version)
- **rating_count_ver** - User Rating counts (for current version)
- **user_rating**  - Average User Rating value (for all version)
- **user_rating_ver** - Average User Rating value (for current version)
- **cont_rating** - Content Rating
- **prime_genre** - Primary Genre
- **sup_devices.num** - Number of supporting devices
- **lang.num** - Number of supported languages

### Google Play dataset:

In [4]:
explore_data(playstore_data, 0, 1, True)

['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']


Number of rows: 10842
Number of columns: 13


In [5]:
explore_data(playstore_data, 1, 5, False)

['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['Coloring book moana', 'ART_AND_DESIGN', '3.9', '967', '14M', '500,000+', 'Free', '0', 'Everyone', 'Art & Design;Pretend Play', 'January 15, 2018', '2.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up']




### Columns which similar content in both datasets

| Description          | Google Play Dataset Column Name | App Store Dataset Column Name |
|----------------------|---------------------------------|-------------------------------|
| Application name     | **App**                         | track_name                    |
| Rating               | **Rating**                      | **user_rating** - for all versions|
|                      |                                 | **user_rating_ver** - for last version |
| Reviews              | **Reviews**                     | **rating_count_tot** - for all versions |
|                      |                                 | **rating_count_ver** - for last version  |
| Application Size     | **Size** - in Megabytes         | **size_bytes** - size in bytes.|
| Is this a free app?  | **Type** - choose "Free"        | **price** - must be equal to 0 |



### Data cleaning

#### Removal of invalid entries

Following [this discussion](https://www.kaggle.com/datasets/lava18/google-play-store-apps/discussion/66015) we remove the row with missing data from the Google Play Data.

In [6]:
cleaned_playstore_data = playstore_data

print(cleaned_playstore_data[10473])

del cleaned_playstore_data[10473]

['Life Made WI-Fi Touchscreen Photo Frame', '1.9', '19', '3.0M', '1,000+', 'Free', '0', 'Everyone', '', 'February 11, 2018', '1.0.19', '4.0 and up']


On the basis of [this discussion](https://www.kaggle.com/datasets/ramamet4/app-store-apple-data-set-10k-apps/discussion?sort=undefined) we do investigate the missing values in the App Store dataset. 

In [7]:
cleaned_appstore_data = appstore_data

incorrect_rows: int = 0
len_header = len(cleaned_appstore_data[0])    
for row in appstore_data[1:]:
    if len(row) != len_header:
        ++incorrect_rows
        
print(f"incorrect_rows: {incorrect_rows}")

incorrect_rows: 0


We haven't found any rows qualifying for the deletion.

### Duplicate entries in the Google Play Dataset

As the code below shows there are duplicated entries for applications within the Google Play Dataset.

In [8]:
all_app_names = dict()
duplicate_app_names = dict()

NAME_IDX = 0
index = 1

def append_to_dict(my_dict, key, index):
    all_ids = my_dict.get(key,[])
    all_ids.append(index)
    my_dict[key] = all_ids
    
for row in cleaned_playstore_data[1:]:
    current_name = row[NAME_IDX]
    if current_name in all_app_names:
        if current_name not in duplicate_app_names:
            append_to_dict(duplicate_app_names, current_name, all_app_names[current_name][0])
        append_to_dict(duplicate_app_names, current_name, index)
    
    append_to_dict(all_app_names, current_name, index)
    index+=1
    
print(f"Duplicate applications with their indices in the dataset: \n {duplicate_app_names}")

Duplicate applications with their indices in the dataset: 
 {'Quick PDF Scanner + OCR FREE': [223, 230, 286], 'Box': [205, 237, 266], 'Google My Business': [194, 240, 269], 'ZOOM Cloud Meetings': [214, 257], 'join.me - Simple Meetings': [254, 262, 283], 'Zenefits': [238, 267], 'Google Ads': [239, 268, 4980], 'Slack': [241, 270, 295], 'FreshBooks Classic': [242, 271], 'Insightly CRM': [243, 272], 'QuickBooks Accounting: Invoicing & Expenses': [244, 273, 297], 'HipChat - Chat Built for Teams': [245, 274], 'Xero Accounting Software': [246, 275], 'MailChimp - Email, Marketing Automation': [247, 276], 'Crew - Free Messaging and Scheduling': [248, 277], 'Asana: organize team projects': [249, 278], 'Google Analytics': [250, 279], 'AdWords Express': [251, 280], 'Accounting App - Zoho Books': [252, 281], 'Invoice & Time Tracking - Zoho': [253, 282], 'Invoice 2go — Professional Invoices and Estimates': [255, 284], 'SignEasy | Sign and Fill PDF and other Documents': [229, 285], 'Genius Scan - PDF

In [9]:
print(f"Names of the duplicated apps: {duplicate_app_names.keys()}")

Names of the duplicated apps: dict_keys(['Quick PDF Scanner + OCR FREE', 'Box', 'Google My Business', 'ZOOM Cloud Meetings', 'join.me - Simple Meetings', 'Zenefits', 'Google Ads', 'Slack', 'FreshBooks Classic', 'Insightly CRM', 'QuickBooks Accounting: Invoicing & Expenses', 'HipChat - Chat Built for Teams', 'Xero Accounting Software', 'MailChimp - Email, Marketing Automation', 'Crew - Free Messaging and Scheduling', 'Asana: organize team projects', 'Google Analytics', 'AdWords Express', 'Accounting App - Zoho Books', 'Invoice & Time Tracking - Zoho', 'Invoice 2go — Professional Invoices and Estimates', 'SignEasy | Sign and Fill PDF and other Documents', 'Genius Scan - PDF Scanner', 'Tiny Scanner - PDF Scanner App', 'Fast Scanner : Free PDF Scan', 'Mobile Doc Scanner (MDScan) Lite', 'TurboScan: scan documents and receipts in PDF', 'Tiny Scanner Pro: PDF Doc Scan', 'Docs To Go™ Free Office Suite', 'OfficeSuite : Free Office + PDF Editor', 'WhatsApp Messenger', 'Messenger – Text and Video

In [10]:
print(f"Number of duplicated app names: {len(duplicate_app_names.keys())}")

Number of duplicated app names: 798


Some of the duplicate entries differ by the number or reviews - in this example these are: 4604324, 4604483, 4604324

In [11]:
for idx in duplicate_app_names['Gmail']:
    print(cleaned_playstore_data[idx])

['Gmail', 'COMMUNICATION', '4.3', '4604324', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Everyone', 'Communication', 'August 2, 2018', 'Varies with device', 'Varies with device']
['Gmail', 'COMMUNICATION', '4.3', '4604483', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Everyone', 'Communication', 'August 2, 2018', 'Varies with device', 'Varies with device']
['Gmail', 'COMMUNICATION', '4.3', '4604324', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Everyone', 'Communication', 'August 2, 2018', 'Varies with device', 'Varies with device']


In other cases the entries are exactly the same:

In [12]:
for idx in duplicate_app_names['Box']:
    print(cleaned_playstore_data[idx])

['Box', 'BUSINESS', '4.2', '159872', 'Varies with device', '10,000,000+', 'Free', '0', 'Everyone', 'Business', 'July 31, 2018', 'Varies with device', 'Varies with device']
['Box', 'BUSINESS', '4.2', '159872', 'Varies with device', '10,000,000+', 'Free', '0', 'Everyone', 'Business', 'July 31, 2018', 'Varies with device', 'Varies with device']
['Box', 'BUSINESS', '4.2', '159872', 'Varies with device', '10,000,000+', 'Free', '0', 'Everyone', 'Business', 'July 31, 2018', 'Varies with device', 'Varies with device']


From the duplicates we will always choose the one with the highest number of Reviews as this data should be the most accurate one.

From the dataset size we have to remove the number of duplicated entries still taking into consideration one entry for each duplicated application.

The expected number of rows in the dataset (after removal of duplicates) is:

In [13]:
 n_row_deduplicated_dataset = len(cleaned_playstore_data[1:]) \
   - sum([len(duplicate_app_names[key]) for key in duplicate_app_names]) \
   + len(duplicate_app_names)
    
 print(f"Number of rows in the deduplicated dataset is: {n_row_deduplicated_dataset}")

Number of rows in the deduplicated dataset is: 9659


In this step we will create dictionary **reviews_max** which will contain the maximal number of reviews for a given application name in the Google Playstore Dataset. We will also verify using the example applications that it is working.

In [14]:
reviews_max = {}

NAME_IDX = 0
REVIEWS_IDX = 3

for row in cleaned_playstore_data[1:]:
    app_name = row[NAME_IDX]
    n_reviews = int(row[REVIEWS_IDX])
    if (app_name not in reviews_max) or (reviews_max[app_name] < n_reviews):
        reviews_max[app_name] = n_reviews
        
print(f"Size of the reviews_max dictionary {len(reviews_max)}")
print("\n--------------------")
print(f"Max reviews for `Box`   = {reviews_max['Box']}")
print(f"Max reviews for `Gmail` = {reviews_max['Gmail']}")

Size of the reviews_max dictionary 9659

--------------------
Max reviews for `Box`   = 159872
Max reviews for `Gmail` = 4604483


In  this step we will create the de-duplicated dataset by choosing the rows with the maximum number of reviews for the given app name. 

The special care must be taken as there can be **multiple rows with the same maximal number of reviews** (this is the case e.g. for the aforementioned **Box** application).

In [15]:
cleaned_playstore_data_v2 = [] 
already_added = set()

# the new dataset should contain the header as well
cleaned_playstore_data_v2.append(cleaned_playstore_data[0])

REVIEWS_ID = 3

for row in cleaned_playstore_data[1:]:
    current_name = row[NAME_IDX]
    current_reviews = int(row[REVIEWS_ID])
    if (current_name not in already_added) and (current_reviews == reviews_max[current_name]): 
        already_added.add(current_name)
        cleaned_playstore_data_v2.append(row)

print("Deduplicated dataset")
print(f"Number of non-header rows: {len(cleaned_playstore_data_v2) - 1}")
print("------------------")
print("First 10 rows of the de-duplicated dataset")
print(f"{cleaned_playstore_data_v2[:10]}")
    

Deduplicated dataset
Number of non-header rows: 9659
------------------
First 10 rows of the de-duplicated dataset
[['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver'], ['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up'], ['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up'], ['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up'], ['Pixel Draw - Number Art Coloring Book', 'ART_AND_DESIGN', '4.3', '967', '2.8M', '100,000+', 'Free', '0', 'Everyone', 'Art & Design;Creativity', 'June 20, 2018', '1.1', '4.4 

### Removal of the non-English applications from both datasets

The cleaned Google Playstore dataset contains the applications which names definitely suggest that they are non-English. 

We do assume that any application which contains at **least 4 characters with the UNICODE value bigger than 127 are non-English** and should be removed from the cleaned version of the dataset.

In [16]:
BOUNDARY_VALUE_ENGLISH_CHARACTERS = 127

def is_english_string(name: str) -> bool:
    non_english_chars_count: int = 0
    for elem in name:
        if ord(elem) > BOUNDARY_VALUE_ENGLISH_CHARACTERS:
            non_english_chars_count += 1
            if non_english_chars_count > 3:
                return False
    return True

print("-----")
for name in ['Instagram', '爱奇艺PPS -《欢乐颂2》电视剧热播', 'Docs To Go™ Free Office Suite',
             'Instachat 😜']:
    print(f"application_name: {name}, is English name = {is_english_string(name)}")

-----
application_name: Instagram, is English name = True
application_name: 爱奇艺PPS -《欢乐颂2》电视剧热播, is English name = False
application_name: Docs To Go™ Free Office Suite, is English name = True
application_name: Instachat 😜, is English name = True


We use the function defined in the preiouvs step to filter out non-English applications from both dataset. For the sake of clarity and brevity **we DO retain the header rows of both datasets**.

Also, please note that we have skipped the v2 version of App Store dataset as it did NOT require any modifications.

In [17]:
#cleaned_playstore_data_v2
#cleaned_appstore_data

def remove_non_english_entries(dataset: list, idx: int):
    result = []
    result.append(dataset[0])
    
    for row in dataset[1:]:
        name = row[idx]
        if is_english_string(name):
            result.append(row)
            
    return result

cleaned_playstore_data_v3 = remove_non_english_entries(cleaned_playstore_data_v2, 0)
cleaned_appstore_data_v3 = remove_non_english_entries(cleaned_appstore_data, 1)

#### Verifying both datasets shape after removal of non-English applications

In [18]:
print("------------")
print(f"Number of non-header rows in the Google Playstore dataset: {len(cleaned_playstore_data_v3) - 1}")
print(f"Number of rows in the App Store dataset: {len(cleaned_appstore_data_v3) - 1}")
print(f"\n First three rows of the Google Playstore dataset:\n\n {cleaned_playstore_data_v3[:3]}")
print(f"\n First three rows of the App Store dataset:\n\n {cleaned_appstore_data_v3[:3]}")

------------
Number of non-header rows in the Google Playstore dataset: 9614
Number of rows in the App Store dataset: 6183

 First three rows of the Google Playstore dataset:

 [['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver'], ['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up'], ['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']]

 First three rows of the App Store dataset:

 [['id', 'track_name', 'size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic'], ['2848822

### Removal of paid applications

From both datasets we are removing paid applications. For the **Google Play Dataset** we execute that by choosing the Type column equal to 'Free'. 

For **App Store Dataset** we do choose applications with price equal to 0.0 (removing $ sign to convert prices to float smoothly).

We did find in the **Google Play Dataset** one row which is inconsistent **('Command & Conquer: Rivals', its type is 'NaN' instead of free, yet its price is '0')**. 

Comparing the size of the inconsistent data with the size of the dataset we can safely omit it from the further analysis. Another argument for that is the fact that that row seems to contain potentially inconsistent data ('NaN' - not-a-number suggests issues with data cleaning or importing or accidental column skipping).

In [19]:
TYPE_COLUMN_IDX = 6
PRICE_COLUMN_IDX = 4

print(f"\nGoogle Playstore name of the column identifying free apps = {cleaned_playstore_data_v3[0][TYPE_COLUMN_IDX]}")
print(f"\nAppStore name of the column identifying free apps = {cleaned_appstore_data_v3[0][PRICE_COLUMN_IDX]}")

cleaned_playstore_data_v4 = []
cleaned_playstore_data_v4.append(cleaned_playstore_data_v3[0])


for row in cleaned_playstore_data_v3[1:]:
    if row[TYPE_COLUMN_IDX].strip().lower() != "free" and row[7] == '0':
        print(f"\nInconsistent row in Google Playstore data is: \n{row}")
        
    if row[TYPE_COLUMN_IDX].strip().lower() == "free":
        cleaned_playstore_data_v4.append(row)
        
cleaned_appstore_data_v4 = []
cleaned_appstore_data_v4.append(cleaned_appstore_data_v3[0])


for row in cleaned_appstore_data_v3[1:]:
    if float(row[PRICE_COLUMN_IDX].strip().replace('$', '')) == 0.0:
        cleaned_appstore_data_v4.append(row)


Google Playstore name of the column identifying free apps = Type

AppStore name of the column identifying free apps = price

Inconsistent row in Google Playstore data is: 
['Command & Conquer: Rivals', 'FAMILY', 'NaN', '0', 'Varies with device', '0', 'NaN', '0', 'Everyone 10+', 'Strategy', 'June 28, 2018', 'Varies with device', 'Varies with device']


We do examine the number of rows remaining in the dataset:

In [20]:
print(f"Size of the final version of the Google Play Dataset (without headers) = {len(cleaned_playstore_data_v4)-1}")
print(f"Size of the final version of the App Store Dataset (without headers) = {len(cleaned_appstore_data_v4)-1}")

Size of the final version of the Google Play Dataset (without headers) = 8863
Size of the final version of the App Store Dataset (without headers) = 3222


### Most common applications by genre

The final goal is to place application in both app stores (Google and Apple). Due to the aforementioned facts we are interested in the successfull **application genres** of these markets.

Naturally, that information is located in different places in both app stores.

In **Google Play Store dataset** the application genre is located in columns: **'Category'** (column index = 1) and **'Genres'** (column index = 9).

In **App Store dataset** the application genre is located in column **'prime_genre'** (column index = 11).

First, we create the frequency tables for the tree columns of our interest. On top of that we do print those tables sorting them descendently by the frequency value.

In [21]:
def freq_table(dataset, index):
    value_count = 0
    
    result: dict[str, int] = dict()
    final_result: dict[str, int] = dict()
        
    for row in dataset:
        value = row[index]
        result[value] = result.get(value,0) + 1
        value_count += 1
        
    for key,value in result.items():
        final_result[key] = 100.0*float(value)/value_count
    
    return final_result

def display_table(dataset, index):
    print("---------------")
    print(f"Values frequencies (in %) for column `{dataset[0][index]}`:\n")
    table = freq_table(dataset[1:], index)
    table_display = []
    for key in table:
        key_val_as_tuple = (table[key], key)
        table_display.append(key_val_as_tuple)

    table_sorted = sorted(table_display, reverse = True)
    for entry in table_sorted:
        print(entry[1], ':', entry[0])


In [22]:
print("Google Playstore data")
CATEGORY_IDX = 1
display_table(cleaned_playstore_data_v4, CATEGORY_IDX)

Google Playstore data
---------------
Values frequencies (in %) for column `Category`:

FAMILY : 18.8987927338373
GAME : 9.725826469592688
TOOLS : 8.462146000225657
BUSINESS : 4.592124562789123
LIFESTYLE : 3.9038700214374367
PRODUCTIVITY : 3.8925871601038025
FINANCE : 3.700778517432021
MEDICAL : 3.5315355974275078
SPORTS : 3.3961412614238973
PERSONALIZATION : 3.317161232088458
COMMUNICATION : 3.2381812027530184
HEALTH_AND_FITNESS : 3.0802211440821394
PHOTOGRAPHY : 2.944826808078529
NEWS_AND_MAGAZINES : 2.798149610741284
SOCIAL : 2.6627552747376733
TRAVEL_AND_LOCAL : 2.335552296062281
SHOPPING : 2.2452894053932075
BOOKS_AND_REFERENCE : 2.1437436533904997
DATING : 1.8616721200496447
VIDEO_PLAYERS : 1.7939749520478394
MAPS_AND_NAVIGATION : 1.399074805370642
FOOD_AND_DRINK : 1.241114746699763
EDUCATION : 1.1621347173643235
ENTERTAINMENT : 0.9590432133589079
LIBRARIES_AND_DEMO : 0.9364774906916394
AUTO_AND_VEHICLES : 0.9251946293580052
HOUSE_AND_HOME : 0.8236488773552973
WEATHER : 0.8010831

In [23]:
GENRES_IDX = 9
display_table(cleaned_playstore_data_v4, GENRES_IDX)

---------------
Values frequencies (in %) for column `Genres`:

Tools : 8.450863138892023
Entertainment : 6.070179397495205
Education : 5.348076272142615
Business : 4.592124562789123
Productivity : 3.8925871601038025
Lifestyle : 3.8925871601038025
Finance : 3.700778517432021
Medical : 3.5315355974275078
Sports : 3.4638384294257025
Personalization : 3.317161232088458
Communication : 3.2381812027530184
Action : 3.102786866749408
Health & Fitness : 3.0802211440821394
Photography : 2.944826808078529
News & Magazines : 2.798149610741284
Social : 2.6627552747376733
Travel & Local : 2.324269434728647
Shopping : 2.2452894053932075
Books & Reference : 2.1437436533904997
Simulation : 2.042197901387792
Dating : 1.8616721200496447
Arcade : 1.8503892587160105
Video Players & Editors : 1.771409229380571
Casual : 1.7601263680469368
Maps & Navigation : 1.399074805370642
Food & Drink : 1.241114746699763
Puzzle : 1.1282861333634209
Racing : 0.9928917973598105
Role Playing : 0.9364774906916394
Libraries 

In [24]:
print("AppStore data")
PRIME_GENRE_IDX = 11
display_table(cleaned_appstore_data_v4, PRIME_GENRE_IDX)

AppStore data
---------------
Values frequencies (in %) for column `prime_genre`:

Games : 58.16263190564867
Entertainment : 7.883302296710118
Photo & Video : 4.9658597144630665
Education : 3.6623215394165114
Social Networking : 3.2898820608317814
Shopping : 2.60707635009311
Utilities : 2.5139664804469275
Sports : 2.1415270018621975
Music : 2.0484171322160147
Health & Fitness : 2.017380509000621
Productivity : 1.7380509000620732
Lifestyle : 1.5828677839851024
News : 1.3345747982619491
Travel : 1.2414649286157666
Finance : 1.1173184357541899
Weather : 0.8690254500310366
Food & Drink : 0.8069522036002483
Reference : 0.5586592178770949
Business : 0.5276225946617008
Book : 0.4345127250155183
Navigation : 0.186219739292365
Medical : 0.186219739292365
Catalogs : 0.12414649286157665


Upon further analysis of the **AppStore dataset / column: prime_genre** we can reach the following conclusions:
- in the analysed dataset of English-language application most of the persented applications have purely entertainment purpose - the most common category are **"Games"** constituting the 58% of presented applications, the 2nd most common category is **"Entertainment"** with presence at the level of 7.88%
- while the third in the popularity ranking category is **"Photo & video"** constituting about 4.97% of the applications, the popularity of the consecutive ranking categories falls below 3%.

It is impossible to express on the basis of this any kind of hard recommendation as the application popularity can but does not have to directly translate to the number of its active users.

The soft recommendation is to take a deeper look at the **Gams / entertainment** category in the further analysis.


Analysing the **Google Play Store dataset** columns: **category** and **genres** we can make the following observations:
- for the **category** column the most popular classes are **family** (18.9%), **Game** (9.72%) and **Tools** (8.46%)
- for the **Genres** column the most popular categories are **Tools** (8.45%), **Entertainment** (6.07%), **Education** (5.34%) and **Business** (4.59%)
- the popularity of the categories/genres for the **Google Play Store dataset** seem to be more evenly distributed among categories without any direct leader taking the majority of the market
- additionally, the popularity ranking shows more potential inclination towards tooling/professional application usage than purely to the entertainment purposes
- similarily to the previous paragraph - recommendation of any of the genres/categories purely on this preliminary frequency analysis is not possible

The **popularity of a particular category/genre** is not defined by the number of applications belonging to it but rather the the average number of installations of those applications. 

For the **App Store** dataset we do not have this data so we have to use the proxy feature - total number of user ratings present **rating_count_tot** column. 

The popularity rating based on that criterion is generated by the following piece of code:

In [25]:
prime_genre_freq_table = freq_table(cleaned_appstore_data_v4[1:], PRIME_GENRE_IDX)

keys = {key for key in prime_genre_freq_table}

result_rating = [()] * len(keys)
RATING_COUNT_TOTAL_IDX = 5

index = 0
for genre in prime_genre_freq_table:
    total = 0
    len_genre = 0
    for row in cleaned_appstore_data_v4[1:]:
        genre_app = row[PRIME_GENRE_IDX]
        if genre_app == genre:
            len_genre += 1
            total += float(row[RATING_COUNT_TOTAL_IDX])

    average_rating_count = total / len_genre
    result_rating[index] = (average_rating_count, genre)
    index += 1
    #print(f"'{genre}'' ==> {average_rating_count}")

result_rating = sorted(result_rating, reverse=True)

print("---------")
print("Genre popularity ranking in AppStore (sorted by the descending popularity):\n")
for elem in result_rating:
    print(f"{elem[1]} - {elem[0]}")
        

---------
Genre popularity ranking in AppStore (sorted by the descending popularity):

Navigation - 86090.33333333333
Reference - 74942.11111111111
Social Networking - 71548.34905660378
Music - 57326.530303030304
Weather - 52279.892857142855
Book - 39758.5
Food & Drink - 33333.92307692308
Finance - 31467.944444444445
Photo & Video - 28441.54375
Travel - 28243.8
Shopping - 26919.690476190477
Health & Fitness - 23298.015384615384
Sports - 23008.898550724636
Games - 22788.6696905016
News - 21248.023255813954
Productivity - 21028.410714285714
Utilities - 18684.456790123455
Lifestyle - 16485.764705882353
Entertainment - 14029.830708661417
Business - 7491.117647058823
Education - 7003.983050847458
Catalogs - 4004.0
Medical - 612.0


The **proposed criterion** for choosing the categories from the **AppStore** is **their popularity**. The **cut-off level** of **28000 ratings** is assumed following which the focus on the following genres is suggested:
1. Navigation
2. Reference
3. Social Networking
4. Music
5. Weather
6. Book
7. Food & dring
8. Finance
9. Photo & Video
10. Travel

For the **Google Playstore** we use the **'Installs'** column as the measure of the application **category** popularity. As it
contains the approximate buckets for the numbers of installs ( e.g. '1,000+') they are treated as the exact numbers during pre-processing step which includes the removal of '+' and ',' signs. 

The results are sorted descending by their popularity.

In [26]:
CATEGORY_IDX = 1
INSTALLS_IDX = 5
categories_table = freq_table(cleaned_playstore_data_v4[1:], CATEGORY_IDX)
all_categories = {key for key in categories_table}
print({key for key in categories_table})

{'MEDICAL', 'ART_AND_DESIGN', 'NEWS_AND_MAGAZINES', 'GAME', 'TOOLS', 'LIBRARIES_AND_DEMO', 'DATING', 'COMMUNICATION', 'PRODUCTIVITY', 'COMICS', 'FOOD_AND_DRINK', 'SHOPPING', 'PARENTING', 'LIFESTYLE', 'PERSONALIZATION', 'EDUCATION', 'PHOTOGRAPHY', 'HOUSE_AND_HOME', 'AUTO_AND_VEHICLES', 'TRAVEL_AND_LOCAL', 'EVENTS', 'SPORTS', 'FAMILY', 'ENTERTAINMENT', 'BEAUTY', 'FINANCE', 'SOCIAL', 'BOOKS_AND_REFERENCE', 'MAPS_AND_NAVIGATION', 'HEALTH_AND_FITNESS', 'WEATHER', 'VIDEO_PLAYERS', 'BUSINESS'}


In [27]:
def format_installs(installs):
    forbidden_chars = {'+',','}
    for current in forbidden_chars:
        installs = installs.replace(current,'')
    return installs

final_ranking = [()]*len(all_categories)
index = 0
for category in all_categories:
    total = 0
    len_category = 0
    for row in cleaned_playstore_data_v4[1:]:
        category_app = row[CATEGORY_IDX]
        if category == category_app:
            installs = float(format_installs(row[INSTALLS_IDX]))
            total += installs
            len_category += 1

    average_installs = total / len_category
    final_ranking[index] = (average_installs, category)
    index += 1

final_ranking = sorted(final_ranking, reverse=True)
print("--------------------")
for current_tuple in final_ranking:
    print(f"{current_tuple[1]} ==> {current_tuple[0]}")
        
        

--------------------
COMMUNICATION ==> 38456119.167247385
VIDEO_PLAYERS ==> 24727872.452830188
SOCIAL ==> 23253652.127118643
PHOTOGRAPHY ==> 17840110.40229885
PRODUCTIVITY ==> 16787331.344927534
GAME ==> 15588015.603248259
TRAVEL_AND_LOCAL ==> 13984077.710144928
ENTERTAINMENT ==> 11640705.88235294
TOOLS ==> 10801391.298666667
NEWS_AND_MAGAZINES ==> 9549178.467741935
BOOKS_AND_REFERENCE ==> 8767811.894736841
SHOPPING ==> 7036877.311557789
PERSONALIZATION ==> 5201482.6122448975
WEATHER ==> 5074486.197183099
HEALTH_AND_FITNESS ==> 4188821.9853479853
MAPS_AND_NAVIGATION ==> 4056941.7741935486
FAMILY ==> 3697848.1731343283
SPORTS ==> 3638640.1428571427
ART_AND_DESIGN ==> 1986335.0877192982
FOOD_AND_DRINK ==> 1924897.7363636363
EDUCATION ==> 1833495.145631068
BUSINESS ==> 1712290.1474201474
LIFESTYLE ==> 1437816.2687861272
FINANCE ==> 1387692.475609756
HOUSE_AND_HOME ==> 1331540.5616438356
DATING ==> 854028.8303030303
COMICS ==> 817657.2727272727
AUTO_AND_VEHICLES ==> 647317.8170731707
LIBRA

During comparison of the ranking lists for **AppStore** and **Google PlayStore** it might be noticed that the following categories have high rankings on both lists (we take the intersection of the top 10s of both lists):
        

| AppStore category | corresponding PlayStore  Category  |
|-------------------|------------------------------------|
| 9. Photo & Video     | 2. VIDEO_PLAYERS                |
| 3.Social Networking | 3. SOCIAL |
| 9.Photo & Video  | 4. PHOTOGRAPHY |
| 10.Travel        | 7. TRAVEL_AND_LOCAL|
 


Those four categories constitute the final recommendation.